### PR TITLE
[ROCm profiler] Add ROCTX annotation support for jax.profiler.TraceAnnotation 

### PR DESCRIPTION
**Pull request by cj401-amd**
For internal review, a better one  than the previous PR https://github.com/ROCm/xla/pull/1044.

# [ROCm profiler] Add ROCTX annotation support for jax.profiler.TraceAnnotation

## Problem

`jax.profiler.TraceAnnotation` on AMD/ROCm hardware produced no visible annotation ranges
in TensorBoard's trace viewer. On CUDA, NVTX ranges from XLA's `ScopedAnnotation` appear
as named bands in `/host:NVTX-CUPTI` and propagate as `kNVTXRange` stats on GPU kernel
events. ROCm had no equivalent pipeline.

## Solution

Four commits implement end-to-end ROCTX annotation support, matching CUDA's NVTX/CUPTI
behavior.

### Commit 1 — Dedicated `/host:ROCTX` XPlane

- Add `kRoctxPlaneName = "/host:ROCTX"` to `xplane_schema.h/.cc`.
- Route `Generic` (ROCTX marker) events from `PerDeviceCollector::Export()` to the new
  plane via an additional `roctx_plane` parameter.
- Merge `/host:ROCTX` into `/host:CPU` in `PostProcessSingleHostXSpace` using the same
  `ChangeOccupiedLineIds` + `MergePlanes` + `RemovePlanes` pattern as NVTX, with line IDs
  starting at `2LL << 32` (NVTX uses `1LL << 32`; host CPU threads use `[0, 1<<32)`).

### Commit 2 — ROCTX label on GPU kernel events as `kNVTXRange`

- Add `roctx_range_map` to `AnnotationMapImpl`; add `LookUpRoctxRange()`.
- Add `GetCurrentRoctxLabel(uint64_t tid) → std::string` to `RocmTracer` (returns by
  value, copy under `roctx_stack_mutex_`).
- Extend the HIP API ENTER callback to read the ROCTX stack top and store it in
  `AnnotationMap` alongside the `AnnotationStack` text; the guard fires when either field
  is non-empty, so standalone ROCTX annotations produce `kNVTXRange` even without a
  concurrent XLA annotation.
- Look up `roctx_range` in `KernelEvent()`, `HipApiEvent()`, and `MemcpyEvent()`. The
  existing `CreateXEvent()` already emits `kNVTXRange` when `event.roctx_range` is
  non-empty.

### Commit 3 — `jax.profiler.TraceAnnotation` emits ROCTX ranges

- Add `PushAnnotation(annotation_name_)` in `TraceMeWrapper` constructor and
  `PopAnnotation()` in `Stop()`.
- Add `~TraceMeWrapper() { Stop(); }` destructor so the push/pop is balanced even when
  the object is used outside a Python context manager.
- The `stopped_` guard makes `Stop()` idempotent.
- Add `@tsl//tsl/profiler/lib:scoped_annotation` BUILD dep.

### Commit 4 — Correctness fixes

- `GetCurrentRoctxLabel`: `absl::string_view` → `std::string` to eliminate dangling
  reference after mutex release. A concurrent `roctxRangePop` can destroy the source
  `RoctxFrame::message` between the lock release and the caller's `Add()` call; the
  value-return eliminates the window entirely.
- `AnnotationMap::Add` early-return: `annotation.empty()` → `annotation.empty() &&
  roctx_range.empty()` so standalone ROCTX annotations (no XLA AnnotationStack text) are
  no longer silently dropped.
- `TraceMeWrapper` destructor added so `PopAnnotation`/`roctxRangePop` is always called.
- Lock ordering comment in `rocm_tracer.h` clarified to document that `Enable()`
  intentionally holds `collector_mutex_` before the roctx locks and explain why this
  reverse nesting is safe.
- `VLOG(3)` in `AnnotationMap::Add` now includes `roctx_range`.
- `is_host_event` guard added to the Generic-event routing in `PerDeviceCollector::Export`
  to document the assumption that ROCTX markers are always host-side.
- Line-ID space partition documented in `post_process_single_host_xplane.cc`.
- Two new tests: `AnnotationMapStoresRoctxRangeWhenAnnotationEmpty` and
  `GetCurrentRoctxLabelReturnedCopyRemainsValidAfterPop`.

## What appears in TensorBoard

| Plane | Events |
|---|---|
| `/host:CPU` | **ROCTX Threads:** `XlaCompile:#module=jit_fn#`, `XlaPass:...`, `Thunk:#hlo_op=...#` (XLA compiler and dispatch via `ScopedAnnotation`) · `train step=N`, `attention_block`, `qk_projection`, `softmax_scale` (user `TraceAnnotation`) |
| `/device:GPU:N` | GPU kernel bars with `kNVTXRange='Thunk:#hlo_op=<op>#'` and `kTfOp='Thunk:'` |

`/host:ROCTX` is correctly absent from the `.pb` — `PostProcessSingleHostXSpace` merges
it into `/host:CPU` at collection time (same as NVTX on CUDA). TensorBoard renders the
merged ROCTX events as **ROCTX Threads** rows within the `/host:CPU` process.

## Tests

- `rocm_tracer_test.cc`: 8 new tests covering push/pop stack semantics,
  `GetCurrentRoctxLabel` lifetime safety (concurrent-pop scenario), `AnnotationMap` roctx
  storage (including the empty-annotation case), and XSpace plane routing (plane-name
  assertions on existing integration tests).
- `roctx_utils_test.cc`: existing 13 tests all pass.
- Manual end-to-end verification:

```
[PASS] StepTraceAnnotation 'train' on /host:CPU           (29 events, avg 721 ms)
[PASS] TraceAnnotation blocks on /host:CPU                (434 events)
[PASS] ROCTX pipeline active — XlaCompile/Thunk visible   (9140 XLA events in /host:CPU)
[PASS] XlaCompile:#module=jit_attention_block# present    (module name in ROCTX event)
[PASS] GPU kernel events carry kNVTXRange  [Commit 2]     (1103 events)
[PASS] GPU kernel events carry kTfOp      [Pipeline A]    (1103 events)
```

## Known limitations

`jax.profiler.TraceAnnotation` labels appear in `/host:CPU` as TraceMe events and in the
ROCTX Threads via `ScopedAnnotation` during JIT compilation, but do **not** propagate to
`kTfOp`/`kNVTXRange` on GPU kernel events during steady-state (post-JIT) execution.

This is an architectural property of JAX's async dispatch: compiled thunks are dispatched
on XLA's executor thread, while `TraceAnnotation` is active on the Python thread. The HIP
API callback fires on the executor thread where the ROCTX stack is empty. During JIT
compilation (the first step when tracing without pre-compilation), labels propagate
correctly because the compiler runs synchronously on the Python thread.

## Files changed

```
xla/backends/profiler/gpu/rocm_collector.cc        — roctx_plane parameter, routing, naming
xla/backends/profiler/gpu/rocm_collector.h         — Export() signature
xla/backends/profiler/gpu/rocm_tracer.cc           — GetCurrentRoctxLabel, HIP callback, event lookups
xla/backends/profiler/gpu/rocm_tracer.h            — GetCurrentRoctxLabel decl, lock comment
xla/backends/profiler/gpu/rocm_tracer_test.cc      — 8 new tests, 2 updated tests
xla/backends/profiler/gpu/rocm_tracer_utils.cc     — AnnotationMap::Add, LookUpRoctxRange, Clear
xla/backends/profiler/gpu/rocm_tracer_utils.h      — roctx_range_map, Add() signature, LookUpRoctxRange
xla/python/BUILD                                   — scoped_annotation dep
xla/python/profiler.cc                             — TraceMeWrapper push/pop/destructor
xla/tsl/profiler/convert/post_process_single_host_xplane.cc  — ROCTX plane merge
xla/tsl/profiler/utils/xplane_schema.cc            — kRoctxPlaneName definition
xla/tsl/profiler/utils/xplane_schema.h             — kRoctxPlaneName declaration
```

### CODE REVIEW

File: xla/tsl/profiler/utils/xplane_schema.cc [hunk 1]
```diff
@@ -53,6 +53,7 @@ TF_CONST_INIT const absl::string_view kCuptiActivityNvtxPlaneName =
     "/host:NVTX-CUPTI";
 TF_CONST_INIT const absl::string_view kRoctracerApiPlaneName =
     "/host:ROCTRACER";
+TF_CONST_INIT const absl::string_view kRoctxPlaneName = "/host:ROCTX";
```
Comment (draganmladjenovic) [inline xla/tsl/profiler/utils/xplane_schema.cc:56]: You should not change the schema. What is the issue using same plane as cuda?

Comment (cj401-amd) [reply xla/tsl/profiler/utils/xplane_schema.cc:56]: I am going to drop the schema change.

**Comment by i-chaochen** [comment]
@cj401-amd do we still need this PR and https://github.com/ROCm/xla/pull/1044 ? if your work are already upstreamed, please close the local ones.
